[doc] Taichi spec: calls, unary ops, binary ops and comparison (#4663)
taichi-dev/taichi: docs/lang/articles/reference.md

#### Calls

> ```

Reviewer: 1. Let's use single space and "" for consistence.
2. We support positional arguments as shown in https://docs.python.org/3/reference/expressions.html#calls, so the rules need to be updated.
3. Our call implementation (https://github.com/taichi-dev/taichi/blob/0e715318cb0b6a19fd2313e930114d63e4960283/python/taichi/lang/ast/ast_transformer.py#L409) includes many cases, so a more detailed description is need:
    - Calling a Python function
    - Calling a Taichi function
    - Calling a Taichi builtin function
    - Calling a type annotation for literals

call          ::=  primary '(' argument_list ')'
argument_list ::=  args
args          ::=  [arg (',' arg)*]
arg           ::=  identifier
```

To favor simplicity, Taichi language doesn't support keyword arguments like Python.

### The power operator

```

> power ::= 'ti.pow(' primary ',' primary ')' | primary ['**' u_expr]

Reviewer: `ti.pow` belongs to Taichi's builtin functions, so it doesn't have to be listed as a syntax rule here.

```

> Taichi predefines its own exponentiation function 'ti.pow()'. Scalars are broadcast in the case of scalar-tensor/tensor-scalar exponentiation operations, and tensor-tensor exponentiation is done elementwise without any broadcasting.

Reviewer: It seems that scalar-tensor/tensor-scalar/tensor-tensor evaluation rules are described multiple times below. Shall we systematically describe it in a new subsection under `Basic concepts`?

Applying Python '**' operator inside Taichi scope is equivalent. In such cases, the power operator binds more tightly than unary operators on the left, but less tightly than unary operators on the right; i.e. -3 ** -2 == -(3 ** (-2)).

### Unary arithmetic and bitwise operations

```
u_expr ::=  power | '-' power | '~' power
```
The unary - operator yields the negation of its argument. The unary ~ operator yields the bitwise inversion of its argument. - can be used with all scalar and tensor. ~ can only be used with integer scalar (i32, i64, etc.) and interrelated tensor.

### Binary arithmetic operations

> ```

Reviewer: Let's add more information regarding type check. See https://github.com/taichi-dev/taichi/blob/0e715318cb0b6a19fd2313e930114d63e4960283/taichi/ir/frontend_ir.cpp#L171

#### Calls

```
call                 ::= primary "(" [argument_list [","] | comprehension] ")"
argument_list        ::= positional_arguments ["," starred_and_keywords]
                           ["," keywords_arguments]
                         | starred_and_keywords ["," keywords_arguments]
                         | keywords_arguments
positional_arguments ::= positional_item ("," positional_item)*
positional_item      ::= assignment_expression | "*" expression
starred_and_keywords ::= ("*" expression | keyword_item)
                         ("," "*" expression | "," keyword_item)*
keywords_arguments   ::= (keyword_item | "**" expression)
                         ("," keyword_item | "," "**" expression)*
keyword_item         ::= identifier "=" expression

```

> To favor simplicity, starred arguments and keyword arguments are not supported for Taichi kernels.
> 
> When calling a Python function, it needs to be wrapped by `ti.static()`.
> 
> The following Taichi builtin functions are supported: `print()`, `min()`, `max()`, `int()`, `float()`, `any()`, `all()`, `abs()`, `pow()`.
> 
> Calling a type annotation for literals

Reviewer: ```suggestion
The `primary` must be evaluated to one of:
- A [Taichi function](basic/syntax.md#taichi-function).
- A [Taichi builtin function](basic/operator.md#other-arithmetic-functions).
- A Taichi primitive type, which serves as a type annotation for a literal. In this case, the `positional_arguments` must be evaluated to a single Python value, and the Python value will be turned into a Taichi value with that annotated type. 
- A Python callable object. If not inside a [static expression](#static-expressions), a warning is produced.
```

The compiler will check both operands to be integral type.

### Binary bitwise operations

```
and_expr ::= shift_expr | and_expr "&" shift_expr
xor_expr ::= and_expr | xor_expr "^" and_expr
or_expr  ::= xor_expr | or_expr "|" xor_expr
```
The & operator computes the bitwise AND of its arguments, the ^ the bitwise XOR, and the | the bitwise OR. The types requirements and broadcast logic are the same as shifting operations.

The compiler will check both operands to be integral type.

### Comparisons

```
comparison    ::= or_expr (comp_operator or_expr)*
comp_operator ::= "<" | ">" | "==" | ">=" | "<=" | "!=" | "is" ["not"] | ["not"] "in"
```
A comparison yields a boolean value (True or False), or if one of the operands is a Tensor, a boolean Tensor. Comparisons can be chained arbitrarily as long as they do not yield boolean Tensors that have more than one element. a op1 b op2 c ... is equivalent to a op1 b and b op2 c and ....

> All operands need to be primaries. The return type is `i32` for all comparision operations.

Reviewer: Comparisons can be chained arbitrarily, e.g., `x < y <= z` is equivalent to `(x < y) & (y <= z)`.

> All operands need to be primaries. The return type is `i32` for all comparision operations.

Reviewer: Comparisons can be chained arbitrarily, e.g., `x < y <= z` is equivalent to `(x < y) & (y <= z)`.

#### Value comparisons

> The operators <, >, ==, >=, <=, and != compare the values of two objects. The two objects generally need to be of the same type, unless there is an implicit type conversion available between the objects. Built-in Python types like strings, lists, tuples are not supported. They need to be converted into Taichi scalar or tensor in advance.

Reviewer: ```suggestion
See [common rules for binary operations](#common-rules-of-binary-operations),
[implicit type casting in binary operations](basic/type.md#implicit-type-casting-in-binary-operations),
and [comparison operators](basic/operator.md#comparison-operators).
```

> The operators <, >, ==, >=, <=, and != compare the values of two objects. The two objects generally need to be of the same type, unless there is an implicit type conversion available between the objects. Built-in Python types like strings, lists, tuples are not supported. They need to be converted into Taichi scalar or tensor in advance.

Reviewer: See [common rules for binary operations](#common-rules-of-binary-operations),
[implicit type casting in binary operations](basic/type.md#implicit-type-casting-in-binary-operations),
and [comparison operators](basic/operator.md#comparison-operators).

#### Membership test operations

> List, dict and tuple types along with 'in' operations, are not supported in Taichi scope. The only exception is to put the whole expression into 'ti.static()' which degenerates into Python expression. More details refer to Python's specifications.

Reviewer: ```suggestion
The semantics of membership test operations follow
[Python](https://docs.python.org/3/reference/expressions.html#membership-test-operations),
but they are only supported in [static expressions](#static-expressions).
```

#### Calls

```

> call                 ::= primary "(" [argument_list [","] | comprehension] ")"
> argument_list        ::= positional_arguments ["," starred_and_keywords]
>                            ["," keywords_arguments]
>                          | starred_and_keywords ["," keywords_arguments]
>                          | keywords_arguments
> positional_arguments ::= positional_item ("," positional_item)*
> positional_item      ::= assignment_expression | "*" expression
> starred_and_keywords ::= ("*" expression | keyword_item)
>                          ("," "*" expression | "," keyword_item)*
> keywords_arguments   ::= (keyword_item | "**" expression)
>                          ("," keyword_item | "," "**" expression)*
> keyword_item         ::= identifier "=" expression
> 
> 
> ```

Reviewer: ```suggestion
call                 ::= primary "(" [positional_arguments] ")"
positional_arguments ::= positional_item ("," positional_item)*
positional_item      ::= assignment_expression | "*" expression
```
